torch_xla_py/xla_dist.py

```diff
@@ -33,17 +33,55 @@ def __init__(self, client_workers, service_workers):
       client_workers: a list of ClientWorker objects.
       service_workers: a list of ServiceWorker objects.
     """
+    for client_worker in client_workers:
+      if not isinstance(client_worker, ClientWorker):
+        raise ValueError(
+            'client_workers argument must be a list of ClientWorker')
+    for service_worker in service_workers:
+      if not isinstance(service_worker, ServiceWorker):
+        raise ValueError(
+            'service_workers argument must be a list of ServiceWorker')
     self._client_workers = list(client_workers)
     self._service_workers = list(service_workers)
 
+
   def validate(self):
     """Validates the current cluster configuration.
 
     Raises:
       RuntimeError: If the cluster is misconfigured, this validation will
-        raise an error.
+        raise an error. For example, if the VMs are in different zones,
+        or not all of the CPU workers have the same size (number of CPU
+        cores, RAM size) we raise an exception. For TPUs we similarly
+        raise an exception if different zones or machine/accelerator_type.
     """
-    raise NotImplementedError()
+    if len(self._client_workers) == 0 or len(self._service_workers) == 0:
+      raise RuntimeError(
+        'Both client_workers and service_workers should not be empty')
+
+    if len(self._client_workers) != len(self._service_workers):
+      raise RuntimeError(
+          'The client_workers and service_workers must have a 1:1 mapping')
+
+    zones = {worker._zone for worker in self._client_workers}
+    zones.update(worker._zone for worker in self._service_workers)
+    if len(zones) != 1:
+      raise RuntimeError(
+          'All workers must be in the same zone, got: {}'.format(zones))
+
+    client_machine_types = {
```

Comment:
What is a Client machine type here?
If it is the VM type, do we want to impose the same VM types?
I am not sure about that.

Reply:
Yes it is the VM type (`n1-standard-16` for example). And yes, I think we should. What if the customer creates VM types of 3 `n1-standard-64` and 1 `n1-standard-1` to feed the v3-32 TPU? It'll severely be bottlenecked by the input pipeline running on `n1-standard-1` silently so we should raise an error before training starts so the user isn't confused why training is so slow.

Reply:
What if the user has just a bunch of user VMs, and even the smaller of it, will be enough to feed his current task?

Reply:
At least it won't be bottlenecked by a single workers due to misconfiguration. I do think we should have same workers on both client and server side?

Reply:
Server side, for sure ;)
Client side, we should give an option to create a client side mesh with non uniform VM set.

Reply:
OK. Can you help me understand why we would only want to raise error for, let's say, 1 `v2-8` and 1 `v3-8` case, but not for 1 `n1-standard-1` and 1 `n1-standard-16` case (or are you saying only raise error when mixing pods with v2/3-8s)? But sure I'm fine with removing the error.

Reply:
I think forcing same TPU HW makes sense (though one might argue that one could override that check as well).
One could have a set of non uniform VMs whose smaller is fast enough to feed the job, and forcing it to create a new set does not make much sense in that case.
So, IMHO, checks are fine by default, but we should provide a way to override.

Reply:
Done. Thanks!